### PR TITLE
DM-31345: Fix the underlying bug resulting in PSF-matching failures in ci_hsc and ci_imsim

### PR DESCRIPTION
**Pull request by arunkannawadi**
The PSF shape needs to be computed at the centroid of the source of interest. There was a subtler issue with the scaling being done on trace radius in some places and on determinant radius in some other places, although this inconsistency had little role to play. Both of these are fixed in this ticket, and the messages about unable to find PSF-matching kernel do not appear any more.

### CODE REVIEW

File: python/lsst/meas/extensions/gaap/_gaussianizePsf.py
Comment (laurenam) [inline python/lsst/meas/extensions/gaap/_gaussianizePsf.py:123]: I can't keep up with the proper use of format strings in logging, but I will note that the use of `f` here does not match the other `self.log.debug` statements I see in this file.

File: python/lsst/meas/extensions/gaap/_gaussianizePsf.py
Comment (laurenam) [inline python/lsst/meas/extensions/gaap/_gaussianizePsf.py:133]: You need to add this new raise to the docs.  Also, the previous behavior did not include a `raise` statement (even though the docs in `_solve()` say that it does)...please give a thought to what the intended behavior is (i.e. should this really be a `raise` or just a logging of the error and move on?)

**Review by laurenam — COMMENTED** [review]
I find the wording around the need to compute the PSF at `center` confusing.   In the commit message you imply wanting it to be the center of the exposure's bounding box, but I think you mean that you want to evaluate the PSF at the centroid of the _source_.  Also, the statement in the PR: 

> The PSF shape is be computed at the centroid of the source being measured, even if the exposure is already clipped to the bounding box of the corresponding footprint

is confusing as what I think you mean to imply is that clipping to the source's footprint bounding box does not guarantee (or even necessarily suggest) that this will be centered on the source's centroid.  Either a clarification of that, or just a simple statement that the PSF needs to be evaluated at the source of interest's centroid (as opposed to whatever default you get without explicitly specifying a position) would be more clear.

Also (could I be more nit-picky!), I find the statement:

> the messages about unable to find PSF-matching kernel do not appear any more

a bit misleading as such a message still _should_ appear in the case of a failure! (i.e. the point here is that you fixed the issue that was causing so many failures, not that you just got rid of a superfluous error message, right?)

File: python/lsst/meas/extensions/gaap/_gaussianizePsf.py
Comment (erykoff) [reply python/lsst/meas/extensions/gaap/_gaussianizePsf.py:123]: For logging, we're not supposed to use `f` strings.  It should be `self.log.debug("Ratio of model to science PSF = %f", fwhmModel/fwhmScience)`

Comment (arunkannawadi) [reply python/lsst/meas/extensions/gaap/_gaussianizePsf.py]: I am also happy to remove this logging from here. It will appear something like: ``psfMatch DEBUG: Ratio of model to science PSF = 1.15``, which may still not be super helpful.

**Comment by arunkannawadi** [comment]
I modified the description in the PR and in the first commit message. Of course, the comment about "Unable to find PSF-matching kernel" is not a generic one but that it does not appear when run on `ci_hsc` and `ci_imsim`.

File: python/lsst/meas/extensions/gaap/_gaussianizePsf.py
Comment (arunkannawadi) [reply python/lsst/meas/extensions/gaap/_gaussianizePsf.py]: The modified logging helped uncover a bug in `meas_base` about how undeblended plugins were initialiized. DM-31442 has to be merged before merging this branch to master.

Comment (laurenam) [reply python/lsst/meas/extensions/gaap/_gaussianizePsf.py]: Oh, right...somehow I missed the `raise e` in the `_solve()` try block...sorry for the noise!  Can you kick off a fresh Jenkins with the new implementation (which I'll have a look at now).

File: python/lsst/meas/extensions/gaap/_gaap.py
Comment (arunkannawadi) [reply python/lsst/meas/extensions/gaap/_gaap.py:281]: The plugin logging framework needs this.
https://pipelines.lsst.io/modules/lsst.meas.base/tasks_and_algorithms.html#how-a-plugin-can-get-its-log-name

Comment (laurenam) [reply python/lsst/meas/extensions/gaap/_gaap.py:281]: Yeah...saw that when the `meas_base` PR popped up!

File: python/lsst/meas/extensions/gaap/_gaussianizePsf.py
Comment (laurenam) [reply python/lsst/meas/extensions/gaap/_gaussianizePsf.py]: Totally your call given that it's at `debug` level 😄

File: tests/test_gaap.py
Comment (laurenam) [inline tests/test_gaap.py:126]: Might be a long-shot...but would it be possible to force a failure to `_solve()` in the test code to make sure it "does the right thing"?

**Review by laurenam — APPROVED** [review]

File: tests/test_gaap.py
Comment (arunkannawadi) [reply tests/test_gaap.py:126]: So, the `testFail` method does that, especially at [L266](https://github.com/lsst/meas_extensions_gaap/blob/cb7397a053f9f69d9587af2d6733b028f990b9e9/tests/test_gaap.py#L266) where `flag_gaussianization` is set to True by the `fail` method after logging. But running pytest or scons doesn't create the log file. Running the test script as `python tests/test_gaap.py` (not recommended and is going to be phased out) prints it on the command line. It looks as follows:

```
.ERROR: Unable to calculate psf matching kernel
measurement.ext_gaap_GaapFlux ERROR: Failed to solve for PSF matching kernel in GAaP for (100.000000, 670.000000): Problematic scaling factors = 100.0 Errors: Exception('Unable to determine kernel sum; 0 candidates')
ERROR: Unable to calculate psf matching kernel
measurement.ext_gaap_GaapFlux ERROR: Failed to solve for PSF matching kernel in GAaP for (100.000000, 870.000000): Problematic scaling factors = 100.0 Errors: Exception('Unable to determine kernel sum; 0 candidates')
ERROR: Unable to calculate psf matching kernel
measurement.ext_gaap_GaapFlux ERROR: Failed to solve for PSF matching kernel in GAaP for (-10.000000, -20.000000): Problematic scaling factors = 100.0 Errors: Exception('Unable to determine kernel sum; 0 candidates')
```

Any inputs on how to run the test_gaap.py in a manner to produce a log file? I was going to run `forcedPhotCoadd` with just the logging changes on the `ci_imsim` and `ci_hsc` branches and see how the error messages look like, but it is of course not principled.

Comment (laurenam) [reply tests/test_gaap.py:126]: Hhhmmm...no, I have no idea how to do that.  It seems there are now no GAaP failures in the limited ci datasets (which is fundamentally a very good thing if the numbers are all sensible!)  But what you have shown here does sufficiently scratch the itch I had to see an example of the log output on error, so thanks!

Comment (timj) [reply tests/test_gaap.py:126]: `.log` seems odd for the name of a logger -- it's not a file name.

File: python/lsst/meas/extensions/gaap/_gaap.py
Comment (timj) [inline python/lsst/meas/extensions/gaap/_gaap.py:337]: No new python code should be using `lsst.log`. `meas_base` does not.  Use `logging.getLogger()`.

What log name does this end up as in practice? Should the `logName` default to something that has `gaap` in it?

File: python/lsst/meas/extensions/gaap/_gaap.py
Comment (timj) [inline python/lsst/meas/extensions/gaap/_gaap.py:336]: You default this to `None` so are you happy for this to be `None`?

File: python/lsst/meas/extensions/gaap/_gaap.py
Comment (timj) [reply python/lsst/meas/extensions/gaap/_gaap.py:281]: That documentation is out of date. It shouldn't be referring to lsst.log in the "how to use" section.

**Review by timj — APPROVED** [review]
This change makes clear the meas_base documentation should have been updated to use python logging.

File: python/lsst/meas/extensions/gaap/_gaap.py
Comment (arunkannawadi) [reply python/lsst/meas/extensions/gaap/_gaap.py:337]: The `BaseMeasurementTask` seems to assign the `logName`, with the plugin name included. https://github.com/lsst/meas_base/blob/c63f6ee5de4496055ae9f4a2d7b9beb5fae67e73/python/lsst/meas/base/baseMeasurement.py#L272

File: python/lsst/meas/extensions/gaap/_gaap.py
Comment (arunkannawadi) [reply python/lsst/meas/extensions/gaap/_gaap.py:336]: The job of providing the `logName` is taken care of by the plugin framework. It actually can never handle `None` case within the plugin, which is why I had to give some string in the unit tests, where I run the plugin outside of the measurement framework. I thought about getting the default logger if `logName` is `None`. Should I go that way?

Comment (timj) [reply python/lsst/meas/extensions/gaap/_gaap.py:336]: Hmm. Okay, it's not clear from the parameter defaulting that this is impossible in general but it sounds like it's common knowledge for all plugin writers. I suppose a comment wouldn't hurt to remind people.

File: python/lsst/meas/extensions/gaap/_gaap.py
Comment (arunkannawadi) [reply python/lsst/meas/extensions/gaap/_gaap.py:337]: The log name will end up as `measurement.ext_gaap_GaapFlux` in practice. Now, unlike `lsst.log`, Python's default logging method doesn't seem to being the message with the logger name. So running the `testFail` method from the command line produces

```
RROR: Unable to calculate psf matching kernel
Failed to solve for PSF matching kernel in GAaP for (100.000000, 670.000000): Problematic scaling factors = 100.0 Errors: Exception('Unable to determine kernel sum; 0 candidates')
ERROR: Unable to calculate psf matching kernel
Failed to solve for PSF matching kernel in GAaP for (100.000000, 870.000000): Problematic scaling factors = 100.0 Errors: Exception('Unable to determine kernel sum; 0 candidates')
ERROR: Unable to calculate psf matching kernel
Failed to solve for PSF matching kernel in GAaP for (-10.000000, -20.000000): Problematic scaling factors = 100.0 Errors: Exception('Unable to determine kernel sum; 0 candidates')
```
unlike the [previous comment](https://github.com/lsst/meas_extensions_gaap/pull/19#discussion_r690851862).

Comment (laurenam) [reply python/lsst/meas/extensions/gaap/_gaap.py:337]: I think that’s actually ok.  See discussion on this PR: https://github.com/lsst/meas_extensions_piff/pull/4

Comment (timj) [reply python/lsst/meas/extensions/gaap/_gaap.py:337]: Yes, this is default python logging and we configure it when pipetask or command line tasks run. If you want to test that a log message is actually being issued you should use `self.assertLogs` and with that you can check that an ERROR is issued and also what the message contains. Testing that an ERROR log message is issued might well be a good idea for this ticket.

File: python/lsst/meas/extensions/gaap/_gaap.py
Comment (arunkannawadi) [reply python/lsst/meas/extensions/gaap/_gaap.py:336]: This is now mentioned in the docstrings, where `logName` is defined.

File: python/lsst/meas/extensions/gaap/_gaap.py
Comment (arunkannawadi) [reply python/lsst/meas/extensions/gaap/_gaap.py:337]: Done.

File: tests/test_gaap.py
Comment (arunkannawadi) [reply tests/test_gaap.py:126]: I've removed specifying any value for `logName`, since the Python logging can handle `None`. The logging aspects themselves are checked in the `testFail` method explicitly.

**Pull request merged**
